### PR TITLE
Feature/appointment

### PR DESCRIPTION
**Pull request by zarmeenmustafa651**

### CODE REVIEW

File: app/views/appointments/_new.html.erb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  <div class="modal-header">
+    <h5 class="modal-title" id="exampleModalLabel">New Appointment</h5>
+    <button type="button" class="close" data-dismiss="modal" aria-label="Close">
+      <span aria-hidden="true">&times;</span>
+    </button>
+  </div>
+  <div class="modal-body">
+      <div class="form-group">
+        <label class="col-form-label">Date: </label> 
+        <%= f.date_field :date, :required => true, class: "form-control" , id: 'date' %>
```
Comment (itratzahra7) [inline app/views/appointments/_new.html.erb]: required: true

File: app/controllers/appointments_controller.rb [hunk 2]
```diff
@@ -86,4 +89,17 @@ def index_page_breadcrumb
   def show_page_breadcrumb
     add_breadcrumb t('appointment.breadcrumb.show'), appointment_path
   end
+
+  def date_to_day(date)
+    if(!date.nil?)
```
Comment (itratzahra7) [inline app/controllers/appointments_controller.rb]: if date.present?

File: app/controllers/appointments_controller.rb [hunk 3]
```diff
@@ -1,9 +1,11 @@
+require 'date'
 class AppointmentsController < ApplicationController
   load_and_authorize_resource find_by: :sequence_num
 
   before_action :index_page_breadcrumb, only: [:index, :new, :show, :edit]
   before_action :show_page_breadcrumb, only: [:show, :edit]
-  
+  helper_method :date_to_day
```
Comment (itratzahra7) [inline app/controllers/appointments_controller.rb]: change this

File: app/controllers/appointments_controller.rb [hunk 4]
```diff
@@ -86,4 +89,17 @@ def index_page_breadcrumb
...
     add_breadcrumb t('appointment.breadcrumb.show'), appointment_path
   end
+
+  def date_to_day(date)
+    if(!date.nil?)
+      date.to_date.strftime("%A")
+    end
+  end
+
+  def show_availabilities
```
Comment (itratzahra7) [inline app/controllers/appointments_controller.rb:94]: resetful comment

File: app/views/appointments/index.html.erb [hunk 5]
```diff
@@ -1,5 +1,49 @@
...
+          </div> 
+        <% end %>  
+      </div>
+    </div> 
+    <div class="card-body">
+      <table  class="table">   
+        <thead >   
+          <tr>
+            <th>Date</th> 
+            <th>Start</th>
```
Comment (itratzahra7) [inline app/views/appointments/index.html.erb]: move to en.yml

File: app/controllers/availabilities_controller.rb [hunk 6]
```diff
@@ -56,4 +56,26 @@ def availability_params
...
+
+  def breakslots
+    weekday = @availability.week_day
+    doctor = @availability.doctor_id
+    check_end = @availability.end_slot
+    starting = @availability.start_slot
+    ending = @availability.start_slot+30.minute
+    @availability.end_slot = @availability.start_slot+30.minute
+    until ending > check_end 
+      if !@availability.save
```
Comment (itratzahra7) [inline app/controllers/availabilities_controller.rb]: unless

File: app/views/appointments/_availabilities_list.html.erb [hunk 7]
```diff
@@ -0,0 +1,2 @@
+<% list = Availability.where(doctor_id: params[:doctor_id], week_day: params[:date]).all %>
```
Comment (itratzahra7) [inline app/views/appointments/_availabilities_list.html.erb]: there shouldn't be any query in views  move this to helper or model

File: app/views/appointments/index.html.erb [hunk 8]
```diff
@@ -1,5 +1,49 @@
...
+        <tbody>   
+        <% @appointments.each do |appointment| %>   
+          <tr> 
+            <td><%= appointment.date %></td>   
+            <% availability = Availability.find(appointment.availability_id) %>
+            <td><%= availability.start_slot.strftime("%I:%M%p") %></td>  
+            <td><%= availability.end_slot.strftime("%I:%M%p") %></td>
+            <% if current_user.type == 'Patient' %>
+              <td><%= Doctor.find(appointment.doctor_id).name %></td> 
+            <% elsif current_user.type == 'Doctor' %>
```
Comment (itratzahra7) [inline app/views/appointments/index.html.erb]: can we make a method in model with name doctor?, patient?

File: app/views/availabilities/_availability_tab.html.erb [hunk 9]
```diff
@@ -3,31 +3,35 @@
...
-              <% if availability.week_day == days[num] %>  
-                <tr>  
-                  <td><%= availability.start_slot.strftime("%I:%M%p") %></td>  
-                  <td><%= availability.end_slot.strftime("%I:%M%p") %></td>
-                  <td>
-                    <%= link_to doctor_availability_path(@doctor, availability), method: :delete do %>
-                      <i class="fas fa-trash-alt"></i>
-                    <% end %>
-                  </td>     
+          <% days = Date::DAYNAMES%>
```
Comment (itratzahra7) [inline app/views/availabilities/_availability_tab.html.erb]: add space

File: app/views/availabilities/_availability_tab.html.erb [hunk 10]
```diff
@@ -3,31 +3,35 @@
...
+          <% if !@availabilities.where(week_day: days[num]).empty? %>
```
Comment (itratzahra7) [inline app/views/availabilities/_availability_tab.html.erb]: make a scope

File: app/views/availabilities/_availability_tab.html.erb [hunk 11]
```diff
@@ -3,31 +3,35 @@
...
+                        <i class="fas fa-trash-alt"></i>
+                      <% end %>
+                    </td>     
+                  </tr>   
+                <% end %>     
+              <% end %>  
+              </tbody>   
+            </table>
+          <% else %>
+            No Availability Slot
```
Comment (itratzahra7) [inline app/views/availabilities/_availability_tab.html.erb]: move to en

File: app/controllers/availabilities_controller.rb [hunk 12]
```diff
@@ -56,4 +56,26 @@ def availability_params
...
     add_breadcrumb t('availability.breadcrumb'), doctor_availabilities_path(@doctor)
   end
+
+  def breakslots
+    weekday = @availability.week_day
+    doctor = @availability.doctor_id
+    check_end = @availability.end_slot
+    starting = @availability.start_slot
+    ending = @availability.start_slot+30.minute
+    @availability.end_slot = @availability.start_slot+30.minute
```
Comment (MirahSaqib) [inline app/controllers/availabilities_controller.rb]: give space befor and after +

File: app/controllers/availabilities_controller.rb [hunk 13]
```diff
@@ -56,4 +56,26 @@ def availability_params
...
+    until ending > check_end 
+      unless @availability.save
```
Comment (MirahSaqib) [inline app/controllers/availabilities_controller.rb]: single line

```return false unless @availability.save```

File: app/controllers/availabilities_controller.rb [hunk 14]
```diff
@@ -56,4 +56,26 @@ def availability_params
   def index_page_breadcrumb
     add_breadcrumb t('availability.breadcrumb'), doctor_availabilities_path(@doctor)
   end
+
+  def breakslots
```
Comment (MirahSaqib) [inline app/controllers/availabilities_controller.rb]: can we move this method to model?

File: app/helpers/availability_helper.rb [hunk 15]
```diff
@@ -0,0 +1,5 @@
+module AvailabilityHelper
+  def availability_list(doctor_id, date)
+    Availability.where(doctor_id: doctor_id, week_day: date).all
```
Comment (MirahSaqib) [inline app/helpers/availability_helper.rb]: pass company too and then use that. No need for .all

```company.availabilities.where(doctor_id: doctor_id, week_day: date)```

File: app/controllers/availabilities_controller.rb [hunk 16]
```diff
@@ -56,4 +56,26 @@ def availability_params
...
+    doctor = @availability.doctor_id
+    check_end = @availability.end_slot
+    starting = @availability.start_slot
+    ending = @availability.start_slot+30.minute
+    @availability.end_slot = @availability.start_slot+30.minute
+    until ending > check_end 
+      unless @availability.save
+        return false
+      end
+      @availability = Availability.new
```
Comment (MirahSaqib) [inline app/controllers/availabilities_controller.rb]: why are we doing this?

File: app/models/ability.rb [hunk 17]
```diff
@@ -18,10 +18,12 @@ def initialize(user)
       can :read, Doctor, hospital_id: user.hospital_id, id: user.id
       can :update, Doctor, hospital_id: user.hospital_id, id: user.id
       can :manage, Availability, hospital_id: user.hospital_id, doctor_id: user.id
+      can :manage, Appointment, hospital_id: user.hospital_id, doctor_id: user.id
     elsif user.patient?
       can :show, Patient, hospital_id: user.hospital_id, id: user.id
       can :update, Patient, hospital_id: user.hospital_id, id: user.id
       can :read, Doctor, hospital_id: user.hospital_id
+      can :manage, Appointment, hospital_id: user.hospital_id, patient_id: user.id
```
Comment (MirahSaqib) [inline app/models/ability.rb]: can a patient approve appointment too?

File: app/javascript/js/common.js [hunk 18]
```diff
@@ -148,8 +148,18 @@ function arrow() {
     });
   }
 }
+function appointment() {
+  
+  $('body').on('click', '.next', function(){
+    var data = { doctor_id: $('#doctor').val(), date: $('#date').val() };
+    $.ajax({ type: "GET", url: '/appointments/show_availabilities', dataType: 'script', data: data });
```
Comment (MirahSaqib) [inline app/javascript/js/common.js]: remove dataType: 'script'

File: app/javascript/js/common.js [hunk 19]
```diff
@@ -148,8 +148,18 @@ function arrow() {
...
+  });
+}
+
 $(document).on('turbolinks:load', arrow)
 $(document).ready(arrow)
+$(document).on('turbolinks:load', appointment)
+$(document).ready(appointment)
```
Comment (MirahSaqib) [inline app/javascript/js/common.js]:
```
$(document).ready(function() {
  collapseSidebar;
  showAvailabilitiesForAppointment;
}
```

File: app/javascript/js/common.js [hunk 20]
```diff
@@ -148,8 +148,18 @@ function arrow() {
     });
   }
 }
+function appointment() {
+
```
Comment (MirahSaqib) [inline app/javascript/js/common.js]: remove extra line

File: app/views/appointments/_appointment_modal.html.erb [hunk 21]
```diff
@@ -0,0 +1,7 @@
+<!-- Modal for appointment -->
+<div class="modal fade" id="exampleModal" tabindex="-1" role="dialog" aria-labelledby="exampleModalLabel" aria-hidden="true">
```
Comment (MirahSaqib) [inline app/views/appointments/_appointment_modal.html.erb]: exampleModal -> appointment_modal

File: app/views/appointments/_new.html.erb [hunk 22]
```diff
@@ -0,0 +1,29 @@
+<%= form_for(@appointment, method: 'post') do |f| %> 
+  <div class="modal-header">
+    <h5 class="modal-title" id="exampleModalLabel">New Appointment</h5>
```
Comment (MirahSaqib) [inline app/views/appointments/_new.html.erb]: internationalization

File: app/views/appointments/index.html.erb [hunk 23]
```diff
@@ -1,5 +1,49 @@
 <div class="container pt-3"> 
   <div class="card">
-  	index
+    <div class="card-header p-3 container-fluid">
+      <div class="row">
+        <h3 class="col-md-10">Appointments</h3>
```
Comment (MirahSaqib) [inline app/views/appointments/index.html.erb]: internationalization

File: app/views/appointments/index.html.erb [hunk 24]
```diff
@@ -1,5 +1,49 @@
...
+        </thead>   
+        <tbody>   
+          <% @appointments.each do |appointment| %>   
+            <tr> 
+              <td><%= appointment.date %></td>   
+              <% availability = Availability.find(appointment.availability_id) %>
+              <td><%= availability.start_slot.strftime("%I:%M%p") %></td>  
+              <td><%= availability.end_slot.strftime("%I:%M%p") %></td>
+              <% if current_user.patient? %>
+                <td><%= Doctor.find(appointment.doctor_id).name %></td>
```
Comment (MirahSaqib) [inline app/views/appointments/index.html.erb]: appointment.doctor.name

File: app/views/appointments/index.html.erb [hunk 25]
```diff
@@ -1,5 +1,49 @@
...
+          <% @appointments.each do |appointment| %>   
+            <tr> 
+              <td><%= appointment.date %></td>   
+              <% availability = Availability.find(appointment.availability_id) %>
+              <td><%= availability.start_slot.strftime("%I:%M%p") %></td>  
+              <td><%= availability.end_slot.strftime("%I:%M%p") %></td>
+              <% if current_user.patient? %>
+                <td><%= Doctor.find(appointment.doctor_id).name %></td> 
+              <% elsif current_user.doctor? %>
+                <td><%= Patient.find(appointment.patient_id).name %></td>
```
Comment (MirahSaqib) [inline app/views/appointments/index.html.erb]: appointment.patient.name

File: app/views/appointments/index.html.erb [hunk 26]
```diff
@@ -1,5 +1,49 @@
...
+              <th><%= t('patient.name') %></th>  
+            <% end %>
+          </tr>   
+        </thead>   
+        <tbody>   
+          <% @appointments.each do |appointment| %>   
+            <tr> 
+              <td><%= appointment.date %></td>   
+              <% availability = Availability.find(appointment.availability_id) %>
+              <td><%= availability.start_slot.strftime("%I:%M%p") %></td>
```
Comment (MirahSaqib) [inline app/views/appointments/index.html.erb]: make method in helper

**Pull request merged**